### DISCUSSION TITLE
Linter para el proyecto

### DISCUSSION THREAD

**Discussion by perealberto** [perealberto/decide-single-zambrano#87] [Ideas]
En relación con este [comentario](https://github.com/3ralon/decide-single-zambrano/pull/80#issuecomment-1849022117) en la última PR de @silvglad. Estoy instalando un linter en el proyecto. Ahora en un rato subiré los cambios.

En principio, el linter actúa antes de hacer commit y utiliza el _black formatter_ de python.

@laurolmer @MarioArocaPaez @silvglad @claugp06

**Comment by MarioArocaPaez** [comment]
Guay ¿Qué tenemos que hacer nosotros para ponernoslo? @3ralon

**Comment by perealberto** [comment]
Buenos días compañeros!

@3ralon @MarioArocaPaez @laurolmer @silvglad @claugp06 

Os voy a explicar cómo funciona el linter del proyecto. He utilizado una librería llamada [**pre-commit**](https://pre-commit.com/). Esta funcionalidad genera, a partir de un archivo YAML de configuración, otro archivo conocido como [**git hook**](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks). Un extracto de la página de git que explica el concepto:

> Como muchos otros VCS, Git tiene su manera de lanzar scripts personalizados cuando ocurren ciertas acciones importantes dentro del proyecto. Existen dos grupos de hooks:
> * Lado del cliente: se activan por operaciones como **_hacer commit_** o **_hacer merge_**.
> * Lado del servidor: se ejecutan por operaciones en la red como **_recibir_** un **_commit pusheado_**.

Luego, con pre-commit, generamos un hook (llamado pre-commit) que se guarda en la ruta `decide-single-zambrano/.git/hooks/pre-commit`. Este se ejecutará siempre antes de hacer un commit (como dice su nombre 😜) y formateará los documentos, eliminando espacios en blanco, comprobando la sintaxis de python, etc. 

Si todo ha ido bien, nos dejará hacer commit sin problemas. Si no, modificará los archivos con "conflictos" y los moverá al área de **_unstagged_**. Revisamos los archivos, hacemos _add_, _commit_ (de nuevo) y _push_.

--------------------------------------

## Cómo configurar un hook con pre-commit

### Instalación de pre-commit

Para instalarnos pre-commit, podemos hacerlo desde el requirements.txt o desde el ppa de python.
```bash
pip install -r requirements.txt
```
```bash
pip install pre-commit
```

### Archivo de configuración del hook

Creamos un archivo de configuración para el hook mediante la sintaxis de YAML, nombrándolo `.pre-commit-config.yaml`. Cada rutina o plugin que se ejecutará comprobando una característica en concreto del archivo se especifica con el atributo **_id_** del YAML.
```yaml
repos:
-   repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v2.3.0
    hooks:
    -   id: check-yaml
    -   id: end-of-file-fixer
    -   id: trailing-whitespace
-   repo: https://github.com/psf/black
    rev: 22.10.0
    hooks:
    -   id: black
```
O bien, generamos un archivo con una configuración básica usando sample-config. Utiliza por defecto **_Black Formatter_**, un linter de python que sigue las reglas de estilado PEP8.
```bash
pre-commit sample-config
```

### Instalar el hook configurado

Para instalarlo, basta usar el comando install, buscará el archivo de configuración en el directorio en el que nos encontremos (gracias al nombre que le hemos puesto) y creará un script de hook para git, guardándolo en el directorio `.git/hooks/`.
```bash
pre-commit install
```
Desde este momento, cada vez que ejecutemos **_git commit_**, pre-commit hará de las suyas, escaneando cada archivo y buscando los posibles errores de estilo. 

### Ejecución en todos los archivos ⚠️

Pre-commit también tiene la funcionalidad de ejecutar este análisis en todos los archivos del proyecto. Sin embargo, hay que tener cuidado, porque seguramente cambie más archivos de los previstos.
```bash
pre-commit run --all-files
```

--------------------------------------

## Enlaces de interés

[1]: [Crear nuevos hooks](https://pre-commit.com/#new-hooks)
[2]: [Documentación de la línea de comandos](https://pre-commit.com/#cli)
[3]: [Hooks y rutinas que soporta pre-commit](https://pre-commit.com/hooks.html)

**Reply by perealberto** [reply]
Así que, en relación con esto. Había pensado que podíamos hacer una rama de limpiado con el linter sobre todo el proyecto para asegurar uniformidad en los estilos y los archivos del repositorio.

@laurolmer @claugp06 @silvglad @MarioArocaPaez

**Reply by laurolmer** [reply]
Dale nomás
